**<center><span style="font-size: 30px;">Проектная работа: Исследование надежности заемщиков</span></center>**

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Ознакомление-с-данными" data-toc-modified-id="Ознакомление-с-данными-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Ознакомление с данными</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li><li><span><a href="#Исследование-данных" data-toc-modified-id="Исследование-данных-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Исследование данных</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-числом-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-числом-детей-и-возвратом-кредита-в-срок?-2.7.1"><span class="toc-item-num">2.7.1&nbsp;&nbsp;</span>Есть ли зависимость между числом детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-2.7.2"><span class="toc-item-num">2.7.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-2.7.3"><span class="toc-item-num">2.7.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-2.7.4"><span class="toc-item-num">2.7.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="Возможные-причины-появления-пропусков-в-исходных-данных.-2.7.5"><span class="toc-item-num">2.7.5&nbsp;&nbsp;</span>Возможные причины появления пропусков в исходных данных.</a></span></li></ul></li><li><span><a href="#Общий-вывод." data-toc-modified-id="Общий-вывод.-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Общий вывод.</a></span></li></ul></li></ul></div>

# Исследование надежности заемщиков


## Ознакомление с данными

In [1]:
import pandas as pd


data = pd.read_csv('/Users/anastasia/Downloads/data.csv')


In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

In [6]:
data['days_employed'] = data['days_employed'].abs()

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть.

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

In [14]:
data['education'] = data['education'].str.lower()

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Исследование данных 

#### Есть ли зависимость между числом детей и возвратом кредита в срок?

In [22]:
def children_grouped(quantity):# Создание функции для категоризации семей по числу детей.
    try:
        if quantity == 0:
            return 'бездетные'
        elif 1 <= quantity <= 2:
            return 'малодетные'
        elif 3 <= quantity <= 5:
            return 'многодетные'
    except:
        pass

In [23]:
data_children_grouped = data.loc[:,['children','debt']]
data_children_grouped['children_category'] = data_children_grouped['children'].apply(children_grouped)
data_children_grouped = data_children_grouped.groupby('children_category').agg({'debt':['count','sum','mean']})
data_children_grouped = data_children_grouped.reset_index().droplevel(0, axis = 1)
data_children_grouped['mean'] = (data_children_grouped['mean']*100).round(2).astype('str') + '%'

In [24]:
data_children_grouped.style.hide_index()

**Вывод:** Клиенты, которые `не имеют детей`, чаще возвращают кредит в долг: доля должников составляет около `7,5%`. Клиенты, у которых `один или два ребенка`, хуже выплачивают кредит: доля невоврата составляет `9,3%`. Для анализа многодетных людей не хватает входных значений, чтобы установить закономерность.

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
def family_grouped(status):# Создание функции для группировки семейного положения клиентов
    if status == 0:
        return 'брак зарегестрирован'
    elif 2 <= status <= 3:
        return 'брак был зарегестрирован'
    else:
        return 'брак не зарегестрирован'

In [26]:
data_family_grouped = data.loc[:,['family_status','family_status_id','debt']]
data_family_grouped['marriage'] = data_family_grouped['family_status_id'].apply(family_grouped)
data_family_grouped = data_family_grouped.groupby(['marriage','family_status']).agg({'debt':['count','sum','mean']})
data_family_grouped = data_family_grouped.reset_index().droplevel(0, axis = 1)
data_family_grouped['mean'] = (data_family_grouped['mean']*100).round(2).astype('str') + '%'

In [27]:
data_family_grouped

count  sum   mean
0  брак был зарегестрирован              в разводе   1189   84  7.06%
1  брак был зарегестрирован         вдовец / вдова    951   63  6.62%
2      брак зарегестрирован        женат / замужем  12261  927  7.56%
3   брак не зарегестрирован  Не женат / не замужем   2796  273  9.76%
4   брак не зарегестрирован       гражданский брак   4134  385  9.31%

**Вывод:** Доля задолжноности `женатых/замужних` клиентов составляет `7,5%`, что ниже доли среди клиентов, состоящих в гражданском браке или `неженатых/незамужних`, для которых данное значение состовляет `9,2%` и `9,7%` соответсвенно. Необходимо отметить, что для анализа взята выборка из 12380 замужних людей, в то время как неженатых значительно меньше – 2813. Полученные значения конверсии долга помогают установить общую тенденцию среди анкетированных: люди, состоящие в зарегестрированном браке или вступашие в него в прошлом, вероятнее погасят кредит в установленный срок.

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [28]:
def income_grouped(income):# Создание функции для группировки клиентов по уровню дохода
    if 0 <= income < 50000:
        return 'низкий уровень дохода'
    elif 50000 <= income <= 200000:
        return 'средний уровень дохода'
    else:
        return 'высокий уровень дохода'

In [29]:
data_income_grouped = data.loc[:,['total_income','debt']]
data_income_grouped['income_class'] = data_income_grouped['total_income'].apply(income_grouped)
data_income_grouped = data_income_grouped.groupby('income_class').agg({'debt':['count','sum','mean']})
data_income_grouped = data_income_grouped.reset_index().droplevel(0, axis = 1)
data_income_grouped['mean'] = (data_income_grouped['mean']*100).round(2).astype('str') + '%'

In [30]:
data_income_grouped.style.hide_index()

**Вывод:** Люди со `средним доходом` (предполагем, что это от 50000 до 200000) чаще задерживают выплату кредита: доля должников составляет `8,5%`. Лучше дела обстоят у людей с `высоким уровнем дохода` (свыше 200000): доля невыплат составляет `7,1%`. Нельзя оценить платежеспособность людей с низким доходом из-за малой выборки. В расчете также не приняты во внимание многие внешние факторы такие как семейное положение, количество детей или место проживания, отражающиеся на денежных возможностях клиента.

#### Как разные цели кредита влияют на его возврат в срок?

In [31]:
data_purpose_grouped = data.groupby('purpose_category').agg({'debt':['count','sum','mean']})
data_purpose_grouped = data_purpose_grouped.reset_index().droplevel(0, axis = 1)
data_purpose_grouped['mean'] = (data_purpose_grouped['mean']*100).round(2).astype('str') + '%'

In [32]:
data_purpose_grouped.style.hide_index()

**Вывод:** Клиенты, оформившие кредит с целью `проведения свадьбы` или `покупки недвижимости`, чаще выплачивают кредит в установленный срок: доля задолжников составляет `7,9%` и `7,2%` соответственно. Доля задолжников среди 
клиентов, собирающихся проводить `операции с автомобилем` или `получать образование`, выше и составляет `9,3%` и `9,2%` соответсвенно. Более высокая платежеспособность среди клиентов, собирающихся приобрести жилье, что объясняется тем, что стоимость недвижимости гораздо выше остальных покупок, требует более стабильного заработка и взвешенного решения для преобретения. Клиент больше заинтересован в своевременных платежах, так как каждый месяц банк начисляет проценты на оставшийся долг,следовательно, сумма долга учеличивается. Например, в случае аннуитентных кредитов, сумма долга растет в геометрической прогрессии.

#### Возможные причины появления пропусков в исходных данных.

*Ответ:* В двух столбцах: **days_employed** и **total_income** отсутствует равное чисто пропущенных значений – `2174`. Вероятнее всего произошла техническая ошибка либо при формировании таблицы с исходными данными для анализа, либо при объединении данных в единый файл, либо при формировании исходной таблицы для анкетирования клиентов, либо по другой причине.

### Общий вывод.

В ходе исследования было изучено влияние различных факторов на платежеспособность клиента. Были найдены следующие закономерности:
- клиенты, чье семейное положение связано с фактом возложения на себя **семейных уз**, с большей вероятностью погасят кредит в срок. Так, доля задолжноности **женатых/замужних** клиентов составляет **7,5%**, что **ниже примерно на 2%** доли неженатых/незамужних** клиентов; 
- кредиторы, взявшие кредит с целью проведения **операций с недвижимостью** или **проведения свадьбы**, также потенциально возращают кредит без отсрочек. Доля должников по сравнению с другими категориями **ниже на 2%** и составляет **7,2%** и **7,9%** соответственно;
- клиенты **без детей** закрывают чаще кредит без долга: доля должников составляет **7,5%**, что **ниже примерно на 1,5%** доли других категорий;
- клиенты из **среднего класса** (с зарабатной платой 50000-200000 рублей) чаще имеют задолжность: доля должников **8,5%**, что **примерно на 1,5% больше** доли других категорий.

Полученные результаты можно будет учитывать при построении модели кредитного скоринга.